languagetool-language-modules/de/src/main/java/org/languagetool/tagging/de/GermanTagger.java

```diff
@@ -1,207 +1,244 @@
-    }
-    return result;
-  }
-
-}
+/* LanguageTool, a natural language style checker 
+ * Copyright (C) 2006 Daniel Naber (http://www.danielnaber.de)
+ * 
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+package org.languagetool.tagging.de;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.regex.Pattern;
+
+import org.jetbrains.annotations.Nullable;
+import org.languagetool.AnalyzedToken;
+import org.languagetool.AnalyzedTokenReadings;
+import org.languagetool.JLanguageTool;
+import org.languagetool.tagging.BaseTagger;
+import org.languagetool.tagging.ManualTagger;
+import org.languagetool.tagging.TaggedWord;
+import org.languagetool.tokenizers.de.GermanCompoundTokenizer;
+import org.languagetool.tools.StringTools;
+
+/**
+ * German part-of-speech tagger, requires data file in <code>de/german.dict</code> in the classpath.
+ * The POS tagset is described in
+ * <a href="https://github.com/languagetool-org/languagetool/blob/master/languagetool-language-modules/de/src/main/resources/org/languagetool/resource/de/tagset.txt">tagset.txt</a>
+ *
+ * @author Marcin Milkowski, Daniel Naber
+ */
+public class GermanTagger extends BaseTagger {
+
+  private final ManualTagger removalTagger;
+  private final Pattern IMPERATIVE_PATTERN = Pattern.compile("[iI](ch|hr)|[eE][rs]|[Ss]ie");
+
+  private GermanCompoundTokenizer compoundTokenizer;
+
+  public GermanTagger() {
+    super("/de/german.dict");
+    try (InputStream stream = JLanguageTool.getDataBroker().getFromResourceDirAsStream(getManualRemovalsFileName())) {
+      removalTagger = new ManualTagger(stream);
+    } catch (IOException e) {
+      throw new RuntimeException("Could not load manual tagger data from " + getManualAdditionsFileName(), e);
+    }
+  }
+
+  //Removes the first part of dash-linked words (SSL-Zertifikat -> Zertifikat)
+  private static String sanitizeWord(String word) {
+    String result = word;
+    String[] splitWord = word.split("-");
+    if (splitWord.length > 1) {
+      for (int i = splitWord.length - 1; i >= 0; i--) {
+        if (!splitWord[i].trim().equals("") && !splitWord[i].trim().equals(null)) {

```

Comment:
I think the 2nd condition is not necessary because `trim` cannot return a null-value. (If it could, the 1st condition would probably throw a NullPointerException. Therefore, I would recommend writing `!"".equals(splitWord[i].trim())`).

Reply:
If splitWord[i] can take null, add `splitWord[i] != null`